src/client.rs

```diff
@@ -0,0 +1,374 @@
+use std::future::Future;
+use std::path::PathBuf;
+
+use anyhow::{anyhow, Error};
+use futures::{channel::mpsc, StreamExt, TryStreamExt};
+use hyper::{
+    body::Bytes,
+    client::{Client, HttpConnector},
+    header::{HeaderValue, AUTHORIZATION, CONTENT_LENGTH},
+    Body, Method, Request, Response, StatusCode, Uri,
+};
+use itertools::Itertools;
+use serde::{
+    de::{Deserialize, Deserializer},
+    ser::{Serialize, Serializer},
+};
+use serde_json::Value;
+
+#[cfg(feature = "compat")]
+use crate::util::compat::StrCompat;
+
+pub const MISC_ERROR_CODE: i64 = -1;
+pub const METHOD_NOT_ALLOWED_ERROR_CODE: i64 = -32604;
+pub const PARSE_ERROR_CODE: i64 = -32700;
+pub const METHOD_NOT_ALLOWED_ERROR_MESSAGE: &'static str = "Method not allowed";
+pub const PRUNE_ERROR_MESSAGE: &'static str = "Block not available (pruned data)";
+
+type HttpClient = Client<HttpConnector>;
+
+#[derive(Debug)]
+pub enum SingleOrBatchRpcRequest {
+    Single(RpcRequest<GenericRpcMethod>),
+    Batch(Vec<RpcRequest<GenericRpcMethod>>),
+}
+impl Serialize for SingleOrBatchRpcRequest {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        match self {
+            SingleOrBatchRpcRequest::Single(s) => s.serialize(serializer),
+            SingleOrBatchRpcRequest::Batch(b) => b.serialize(serializer),
+        }
+    }
+}
+impl<'de> Deserialize<'de> for SingleOrBatchRpcRequest {
```

Comment:
Untagged simply tries deserializing each in order, failing over. Also, if none of the variants match, the error message is useless. This method is more code but it's more efficient, since it decides which variant to deserialize into based on whether it's an object or array.

Reply:
Makes sense